### PR TITLE
Add java implementation to centroid & dp on ranges pages

### PR DESCRIPTION
**Pull request by TruaShamu**

### CODE REVIEW

File: content/4_Gold/DP_Ranges.mdx [hunk 1]
```diff
@@ -97,6 +97,41 @@ int main() {
 
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class jazz {
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Ranges.mdx:106]:
```suggestion
public class Jazz {
```

File: content/4_Gold/DP_Ranges.mdx [hunk 2]
```diff
@@ -97,6 +97,41 @@ int main() {
...
+public static int MAXN = 505;
+public static void main(String[] args) throws IOException {
+Kattio io = new Kattio();
+char[] A = io.next().toCharArray();
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Ranges.mdx:110]: is it just me or is the indentaion kinda messed up here

Comment (SansPapyrus683) [reply content/4_Gold/DP_Ranges.mdx:110]: also i'd love it if you gave a better variable name

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Centroid.mdx [hunk 3]
```diff
@@ -86,6 +90,70 @@ int main() {
...
+public class cses2709 {
+	public static int[] subSize;
+	public static ArrayList<Integer>[] adj;
+	public static int N;
+
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+		N = io.nextInt();
+		adj = new ArrayList[N];
+		for (int i=0; i< N; i++) {
```
Comment (dongliu0426) [inline content/5_Plat/Centroid.mdx:108]:
```suggestion
		for (int i = 0; i < N; i++) {
```

File: content/5_Plat/Centroid.mdx [hunk 4]
```diff
@@ -86,6 +90,70 @@ int main() {
...
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class cses2709 {
+	public static int[] subSize;
+	public static ArrayList<Integer>[] adj;
```
Comment (dongliu0426) [inline content/5_Plat/Centroid.mdx:101]:
```suggestion
	public static List<Integer>[] adj;
```

File: content/5_Plat/Centroid.mdx [hunk 5]
```diff
@@ -86,6 +90,70 @@ int main() {
...
+	public static int N;
+
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+		N = io.nextInt();
+		adj = new ArrayList[N];
```
Comment (dongliu0426) [inline content/5_Plat/Centroid.mdx:107]:
```suggestion
		adj = new List[N];
```

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

**Review by dongliu0426 — APPROVED** [review]
nice

File: content/4_Gold/DP_Ranges.mdx [hunk 6]
```diff
@@ -97,6 +97,41 @@ int main() {
...
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Jazz {
+	public static int MAXN = 505;
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Ranges.mdx:107]:
```suggestion
	public static int MAXN = 500;
```

File: content/4_Gold/DP_Ranges.mdx [hunk 7]
```diff
@@ -97,6 +97,41 @@ int main() {
...
+		int[][] dp = new int[MAXN][MAXN];
+		for (int j = 0; j <= inp.length; j++) {
+			for (int i = 0; i < inp.length - j; i++) {
+				// Base case: duplicate the ith letter.
+				dp[i][i + j] = dp[i + 1][i + j] + 1;
+
+				for (int k = i + 1; k <= (i + j); k++) {
+					if (inp[k] == inp[i]) { // Try to match letters i and k.
+						dp[i][i + j] = Integer.min(dp[i][i + j],
+								dp[i + 1][k - 1] + dp[k + 1][i + j]);
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Ranges.mdx:122]:
```suggestion
						dp[i][i + j] = Math.min(
							dp[i][i + j],
							dp[i + 1][k - 1] + dp[k + 1][i + j]
						);
```

File: content/4_Gold/DP_Ranges.mdx [hunk 8]
```diff
@@ -97,6 +97,43 @@ int main() {
...
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Jazz {
+	public static int MAXN = 500;
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio();
```
Comment (maggie-j-liu) [inline content/4_Gold/DP_Ranges.mdx:109]: add kattio codesnip

File: content/4_Gold/DP_Ranges.mdx [hunk 9]
```diff
@@ -97,6 +97,43 @@ int main() {
...
+		char[] inp = io.next().toCharArray();
+
+		// DP[i][j] is the min number of additions to get "jazz" from index i to j
+		int[][] dp = new int[MAXN][MAXN];
+		for (int j = 0; j <= inp.length; j++) {
+			for (int i = 0; i < inp.length - j; i++) {
+				// Base case: duplicate the ith letter.
+				dp[i][i + j] = dp[i + 1][i + j] + 1;
+
+				for (int k = i + 1; k <= (i + j); k++) {
```
Comment (maggie-j-liu) [inline content/4_Gold/DP_Ranges.mdx:119]:
```suggestion
				for (int k = i + 1; k <= i + j; k++) {
```

File: content/5_Plat/Centroid.mdx [hunk 10]
```diff
@@ -86,6 +90,70 @@ int main() {
...
+	}
+
+	// Find the centroid of the tree (the subtree with <= N/2 nodes)
+	public static int getCentroid(int node, int par) {
+		for (int next : adj[node]) {
+			if (next == par) {
+				continue;
+			}
+
+			if ((subSize[next] * 2) > N) {
```
Comment (maggie-j-liu) [inline content/5_Plat/Centroid.mdx:145]:
```suggestion
			if (subSize[next] * 2 > N) {
```

File: content/4_Gold/DP_Ranges.mdx (hunk 8 above)
Comment (ryanchou-dev) [inline content/4_Gold/DP_Ranges.mdx:109]: did you forget to codesnip kattio 🤔

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]

File: content/5_Plat/Centroid.mdx [hunk 11]
```diff
@@ -86,6 +90,70 @@ int main() {
 
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class cses2709 {
```
Comment (SansPapyrus683) [inline content/5_Plat/Centroid.mdx:99]: better name?

File: content/5_Plat/Centroid.mdx (hunk 10 above)
Comment (SansPapyrus683) [inline content/5_Plat/Centroid.mdx:145]: some comments would be really cool here? idk maybe this is supposed to be prior knowledge relative to the problem but

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by TruaShamu** [comment]
Alright, I added the codesnippet and also updated the comments on the centroid problem.

File: content/5_Plat/Centroid.mdx [hunk 12]
```diff
@@ -86,6 +90,70 @@ int main() {
 
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class findCentroid {
```
Comment (SansPapyrus683) [inline content/5_Plat/Centroid.mdx:99]:
```suggestion
public class FindCentroid {
```

File: content/5_Plat/Centroid.mdx [hunk 13]
```diff
@@ -86,6 +90,70 @@ int main() {
...
+				continue;
+			}
+			// Keep searching for the centroid if there are subtrees with more than N/2 nodes.
+			if (subSize[next] * 2 > N) {
+				return getCentroid(next, node);
+			}
+		}
+		return node;
+	}
+	// Codesnip{Kattio}
```
Comment (SansPapyrus683) [inline content/5_Plat/Centroid.mdx:151]: you sure this works? i thought it was `//CodeSnip{Kattio}` w/o the space in the middle

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
this should be it

**Review by SansPapyrus683 — APPROVED** [review]
image

**Comment by TruaShamu** [comment]
😂🤣

File: content/4_Gold/DP_Ranges.mdx [hunk 14]
```diff
@@ -97,6 +97,44 @@ int main() {
...
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Jazz {
+	public static int MAXN = 500;
```
Comment (nathangong) [inline content/4_Gold/DP_Ranges.mdx:107]:
```suggestion
	public static final int MAXN = 500;
```

**Review by nathangong — APPROVED** [review]

**Review by maggie-j-liu — APPROVED** [review]

**Pull request merged**